src/decoder/mod.rs

```diff
@@ -167,16 +167,8 @@ fn decode_array<R: Read + ?Sized>(reader: &mut R, utf8_lossy: bool) -> DecoderRe
             break;
         }
 
-        // check that the key is as expected
-        let key = read_cstring(reader)?;
-        match key.parse::<usize>() {
-            Err(..) => return Err(DecoderError::InvalidArrayKey(arr.len(), key)),
-            Ok(idx) => {
-                if idx != arr.len() {
-                    return Err(DecoderError::InvalidArrayKey(arr.len(), key));
-                }
-            }
-        }
+        // read key, but do no validate value
+        read_cstring(reader)?;
```

Comment:
The corpus tests that "degenerate" BSON where the keys of an array document are not increasing integers is still parsed correctly, so I removed this validation section.